language/move-compiler/src/to_bytecode/canonicalize_handles.rs

```diff
@@ -0,0 +1,466 @@
+/// defined in a module can be arranged in definition order.
+fn struct_definition_order(
+    defs: &[StructDefinition],
+) -> HashMap<StructHandleIndex, StructDefinitionIndex> {
+    defs.iter()
+        .enumerate()
+        .map(|(ix, def)| (def.struct_handle, StructDefinitionIndex(ix as TableIndex)))
+        .collect()
+}
+
+/// Reverses mapping from `FunctionDefinition(Index)` to `FunctionHandle`, so that handles for
+/// structs defined in a module can be arranged in definition order.
+fn function_definition_order(
+    defs: &[FunctionDefinition],
+) -> HashMap<FunctionHandleIndex, FunctionDefinitionIndex> {
+    defs.iter()
+        .enumerate()
+        .map(|(ix, def)| (def.function, FunctionDefinitionIndex(ix as TableIndex)))
+        .collect()
+}
+
+/// Update references to `StructHandle`s within signatures according to the permutation defined by
+/// `structs`.
+fn permute_signature_token(token: &mut SignatureToken, structs: &[TableIndex]) {
+    use SignatureToken as T;
+    match token {
+        T::Bool
+        | T::U8
+        | T::U16
+        | T::U32
+        | T::U64
+        | T::U128
+        | T::U256
+        | T::Address
+        | T::Signer
+        | T::TypeParameter(_) => (),
+
+        T::Vector(token) | T::Reference(token) | T::MutableReference(token) => {
+            permute_signature_token(token, structs)
+        }
+
+        T::Struct(handle) => permute!(StructHandleIndex, *handle, structs),
+
+        T::StructInstantiation(handle, tokens) => {
+            permute!(StructHandleIndex, *handle, structs);
+            for token in tokens {
+                permute_signature_token(token, structs)
+            }
+        }
+    }
+}
+
+/// Update references to function handles within code according to the permutation defined by
+/// `functions`.
+fn permute_code(code: &mut CodeUnit, functions: &[TableIndex]) {
+    for instr in &mut code.code {
+        if let Bytecode::Call(function) = instr {
+            permute!(FunctionHandleIndex, *function, functions);
+        }
+    }
+}
+
+/// Calculates the permutation of indices in `pool` that sorts it according to the key function
+/// `key`:  The resulting `permutation` array is such that, new `pool'` defined by:
+///
+///   pool'[permutation[i]] = pool[i]
+///
+/// is sorted according to `key`.
+fn permutation<'p, T, K: Ord>(
+    pool: &'p Vec<T>,
+    key: impl Fn(TableIndex, &'p T) -> K + 'p,
+) -> Vec<TableIndex> {
+    let mut inverse: Vec<_> = (0..pool.len() as TableIndex).collect();
+    inverse.sort_by_key(move |ix| key(*ix, &pool[*ix as usize]));
+
+    let mut permutation = vec![0 as TableIndex; pool.len()];
+    for (ix, jx) in inverse.into_iter().enumerate() {
+        permutation[jx as usize] = ix as TableIndex;
+    }
+
+    permutation
+}
+
+/// Re-order `pool` according to the `permutation` array.  `permutation[i]` is the new location of
+/// `pool[i].
+fn apply_permutation<T>(pool: &mut Vec<T>, mut permutation: Vec<TableIndex>) {
```

Comment:
It took me a little bit realize how this works lol, feels like a tortuous interview problem

Reply:
But its a bit fun :) I only really understood how it worked when I sat down and said "how would I write this" and you start with the outer loop and quickly realize this is the only thing to do

Reply:
Stack Overflow suggests a different flavour of this algorithm: https://stackoverflow.com/questions/16501424/algorithm-to-apply-permutation-in-constant-memory-space

But, I found it easier to reason about this one.  It does have a flavour of an interview question -- when I did coding interview, I used to ask "Graph Clone", which is not too different from what we're doing here, except that we're cloning the graph onto itself!